[16974689] Skeleton code template - project structure only

aeshanw-aelf opened the pull request:
- skeleton code template for the contract_plugin
- based off the top-level hierrachy of C++ example --> 
   - starting from https://github.com/AElfProject/contract-plugin/blob/master/src/contract_csharp_generator.h#L39
   -  then proceeding down to the top-level sub-functions
     - https://github.com/AElfProject/contract-plugin/blob/master/src/contract_csharp_generator.cc#L612
- just to note im having trouble converting FileDescriptorProto --> FileDescriptor (which is typically used in the C++ example) perhaps due to my lack of exp with Reflections tho open to any suggestions anyone might have 🙇  
 - I had a workaround using `FileDescriptor.BuildFromByteStrings()` --> https://github.com/aeshanw-aelf/contract_plugin-csharp/pull/1/files#diff-e5b527d55deb2d3dc25ca6897248fdd58e3cac56a0e956384e076f93758b3e23R48

aeshanw-aelf committed:
[16974689] Skeleton code template

aeshanw-aelf committed:
solution for FileDescriptor from stdin

aeshanw-aelf committed:
some extra handling for Comment Proto GetSourceLocation

aeshanw-aelf committed:
review-comments: include C++ original code links per unimplemented funcs.throw exception for the same

gldeng commented on ContractPlugin/CsharpComments.cs:
```diff
@@ -0,0 +1,13 @@
+using Google.Protobuf.Reflection;
+
+namespace contract_plugin_csharp;
```
Please follow C# coding style https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names#naming-conventions .

gldeng commented on ContractPlugin/Program.cs:
Let's separate the exe and lib into to projects and make exe reference the lib. The lib can be re-used by other programs.

aeshanw-aelf replied:
hmm ive seperated out the generator-main logic to `ContractGenerator` to serve as the lib. hence Program is alot smaller. just confirm this is the right approach before I move Program.cs (executable) into a seperate project?

gldeng commented on ContractPlugin/CsharpContainer.cs:
```diff
@@ -0,0 +1,81 @@
+namespace contract_plugin_csharp;
+
+// Generates the overall "container" for the generated C# contract
+public class CSharpContainer
```
Let's make the name more meaningful. It's a generator for ContractContainer, we can call it `ContractContainerGenerator`. We can define an interface for `IGenerator<T>` and make T the input type of the descriptor.

aeshanw-aelf committed:
review-comments: some renaming as per C# convention. attempt to use a generic-interface

aeshanw-aelf committed:
rename files to match Class-names. Interface to handle 1 & 2 types

aeshanw-aelf committed:
seperate out lib-logic to ContractGenerator class from ProgramMain

aeshanw-aelf commented:
Hi @gldeng i did a refactor based on your comments can help re-review and just check if im on the right path. thx!

aeshanw-aelf committed:
refactor based on PR comments. Libarize-ContractGenerator. Follow a more static function style

aeshanw-aelf committed:
some extra method changes

aeshanw-aelf committed:
added more comments to unimplemented methods

aeshanw-aelf committed:
include ProtoUtil for unavailable C++ methods

aeshanw-aelf committed:
exclude bin/obj folders

aeshanw-aelf committed:
include READMEs

aeshanw-aelf commented:
Hi @gldeng as discussed i've refactored the code
- broken up into Lib & exe projects (updated soln accordingly)
- Included <xml> comments to methods
- added a skeleton of the C++ specific lib methods in ProtoUtils.cs (will implement in parallel as you review the rest of the structure).

aeshanw-aelf committed:
fix GetFileNamespace signature as per C++

aeshanw-aelf committed:
include missing util methods from C++

aeshanw-aelf committed:
add missing extra event comments

aeshanw-aelf committed:
protoc-compile aelf-options for use by Event low-level methods

aeshanw-aelf committed:
remove all biz-logic and make this a pure project-skeleton PR

aeshanw-aelf commented:
As requested by @gldeng  ive stripped this down to a pure project structure skeleton for ease of PR review.

aeshanw-aelf commented:
As discussed ive moved this review into 3 broken down PRs
- https://github.com/aeshanw-aelf/contract_plugin-csharp/pull/3
- https://github.com/aeshanw-aelf/contract_plugin-csharp/pull/4
- https://github.com/aeshanw-aelf/contract_plugin-csharp/pull/5

hence closing this one
